`project destroy`

paladin-devops opened the pull request:
Adds a new command to Waypoint for destroying a project. By default, this command will destroy all resources created for apps within a project, by initiating a destroy operation for each app in a project. After all resources are destroyed, the project is deleted from the database. The user may opt-out of destroying resources with the `-skip-destroy-resources` flag. Confirmation with either the `-auto-approve` flag, or interactive input is required in order to run the command.

Closes #1418.
Closes #3824.

paladin-devops committed:
Add ProjectDestroy RPC with request and response messages to server protobuf.

paladin-devops committed:
Add ProjectDestroy RPC with request and response messages to server protobuf.

paladin-devops committed:
Add ProjectState field to Project protobuf.

The new ProjectState field is used to signify if a project is active or is in the process of being destroyed.

paladin-devops committed:
Add project destroy operation to server proto for runners to run project destroy op.

paladin-devops committed:
Update server proto to include DestroyProjectOp job type. Update Mockery version for generated files.

The DestroyProjectReponse message is removed, and the DestroyProject RPC now returns an empty protobuf message. DestroyProjectRequest is updated to no longer include the option for destroying resources, which has instead been moved to a new field in DestroyProjectOp.

paladin-devops committed:
Update server proto.

The DestroyProjectReponse message is removed, and the DestroyProject RPC now returns an empty protobuf message. DestroyProjectRequest is updated to no longer include the option for destroying resources, which has instead been moved to a new field in DestroyProjectOp.

paladin-devops committed:
Add an operation to the Waypoint client to destroy a project.

paladin-devops committed:
Add validation for a project destroy request.

paladin-devops committed:
Fix destroy project operation execution.

Panic if the operation type is incorrect for destroying a project. Set the project name for the upsert request to update the project's state.

paladin-devops committed:
Add project destroy command.

paladin-devops committed:
Remove log param from executeDestroyProjectOp.

paladin-devops committed:
Project destroy command docs.

briancain commented on internal/cli/project_destroy.go:
```diff
@@ -0,0 +1,110 @@
+		// Confirmation is required for destroying a project &/or its resources
+		if !c.confirm {
+			proceed, err := c.ui.Input(&terminal.Input{
+				Prompt: "Do you really want to destroy this project and its resources? Only 'yes' will be accepted to approve: ",
```
It'd be nice if we formatted this to say the specific project like

> "Do you really want to destroy project "my-project" and its resources? ....

briancain replied:
It'd be nice if we said the project here just for extra confirmation:

> "Do you really want to destroy project "my-project" and its resources? ...

briancain commented on internal/cli/project_destroy.go:
```diff
@@ -0,0 +1,110 @@
+					terminal.WithErrorStyle(),
+				)
+			} else if strings.ToLower(proceed) != "yes" {
+				app.UI.Output("Destroying project and resources requires confirmation.", terminal.WithWarningStyle())
```
Maybe something like

> Destroying project %q and resources....

briancain commented on internal/cli/project_destroy.go:
```diff
@@ -0,0 +1,110 @@
+
+		f.BoolVar(&flag.BoolVar{
+			Name:    "auto-approve",
+			Usage:   "Auto-approve destroying the project & all resources. If unset, confirmation will be requested.",
```
> "Destroy the project and all of its resources without confirmation"

briancain commented on internal/cli/project_destroy.go:
```diff
@@ -0,0 +1,110 @@
+}
+
+func (c *ProjectDestroyCommand) Synopsis() string {
+	return "Delete the specified project."
```
```suggestion
	return "Delete the specified project and optionally destroy its resources."
```

briancain commented on internal/cli/project_destroy.go:
```diff
@@ -0,0 +1,110 @@
+  You can optionally skip destroying the resources by setting
+  -skip-destroy-resources to true.
+
+  The workspace can continue to be used after this call, this just deletes
```
I'm curious what it means to continue to use the workspace after? Is there another way to say this, I'm not sure what workspace we're referring to.

paladin-devops replied:
I'm removing this for now, we want the operation to destroy all the resources for apps in a project in all workspaces where they're deployed.

briancain commented on internal/runner/operation_project_destroy.go:
```diff
@@ -0,0 +1,71 @@
+	}
+
+	// Update the project state to indicate that it's being destroyed
+	// TODO: Do we need to do something with the response here?
```
Probably not, UpsertProject will return the same Project proto message that was within the request i.e. the value you set within `&pb.UpsertProjectRequest{Project: ... }`

https://github.com/hashicorp/waypoint/blob/d0b40daa74d483a3a8838e43f93bfada5736270d/pkg/server/singleprocess/service_project.go#L20-L42

briancain commented on internal/runner/operation_project_destroy.go:
```diff
@@ -0,0 +1,71 @@
+	if !destroyProjectOp.DestroyProject.SkipDestroyResources {
+		for _, app := range project.Apps() {
+			// TODO: Do we need to do something with the response here?
+			_, err := r.executeDestroyOp(ctx,
```
I don't think so, looks like this returns an empty `pb.Job_Result{}`

https://github.com/hashicorp/waypoint/blob/25b2130a43e4594e67c897884f6934aa0f19d9b9/internal/runner/operation_destroy.go#L42

briancain commented on pkg/server/proto/server.proto:
```diff
@@ -2280,6 +2290,11 @@ message Job {
 
+  message DestroyProjectOp {
+    Project project = 1;
+    bool skip_destroy_resources = 2;
```
I recommend adding some doc comments here, our job operations are pretty well documented.

briancain commented on pkg/server/singleprocess/service_project.go:
```diff
@@ -79,6 +79,23 @@ func (s *Service) ListProjects(
+	}
+
+	// TODO: Do we need to return something other than empty?
+	return &empty.Empty{}, nil
```
In this case, maybe not? Where do we queue the DestroyOperation? In that case we should return the job ID associated with destroying the resources. But since it looks like this simply deletes Project from the DB, it's safe to return empty.

briancain reviewed (commented):
Looking pretty solid @paladin-devops ! I know it's a draft but I had some answers for TODOs and some feedback on user-facing strings that I figured I'd review. Excited for this one!! 👏🏻

paladin-devops committed:
Update messages to user during project destroy & command info.

paladin-devops committed:
Clean up TODOs. Add comments re: destroying projects in protobuf.

paladin-devops committed:
Add log to destroy project operation. Loop through workspaces during destruction of resources.

paladin-devops committed:
Add DestroyProject type to output returned by job commands.

paladin-devops committed:
Update help of project destroy.

paladin-devops committed:
Update project destroy operation to return a job ID in the result.

paladin-devops committed:
Changelog.

paladin-devops committed:
Use Waypoint SDK version with changed to support reporting resources on Destroy operations.

paladin-devops committed:
Update server proto to include DestroyedResources message, and add a field of that type to the Deployment message.

paladin-devops committed:
Set destroyed and remaining resources after a destroy op runs.

The proto message used in the execution of a deployDestroyOperation now sets two fields of a Deployment proto message, DeclaredResources and DestroyedResources. These are available to the caller that initiated the operation with doOperation. The data in these fields is used to output the resources that were and were not destroyed during the destroy operation.

paladin-devops committed:
Updates resourceManager for ECS platform plugin to set destroyed and declared resources during a destroy op.

cicoyle committed:
moved DestroyProject to be a project method and pulled logic out of DoApp, since DestroyProject is at the project level, and not app level.

paladin-devops committed:
Consolidate output messages of destroyed and remaining resources after a destroy op.

paladin-devops committed:
Return an error if the proto message set during a destroy deployment operation can't be converted to a Deployment.

paladin-devops committed:
Regenerate proto.

paladin-devops committed:
gofmt.

paladin-devops committed:
Regenerate server proto.

paladin-devops committed:
Check if there are any destroyed or remaining resources before writing output.

paladin-devops committed:
Tweak output of list of destroyed and remaining resources.

xiaolin-ninja commented on internal/cli/project_destroy.go:
```diff
@@ -0,0 +1,100 @@
+	return formatHelp(`
+Usage: waypoint project destroy [options]
+
+  Delete the project and all resources created for all apps within the project.
```
We could be more descriptive about what these resources are. e.g. "local and network resources" to be extra clear for users.

xiaolin-ninja replied:
I would expand on what type of resources these are to be extra clear for the user. Just Waypoint resources? Plugin resources as well?

catsby commented on internal/cli/project_destroy.go:
```diff
@@ -0,0 +1,100 @@
+			Style:  "",
+			Secret: false,
+		})
+		if err != nil {
+			c.ui.Output(
+				"Error getting input: %s",
+				clierrors.Humanize(err),
+				terminal.WithErrorStyle(),
+			)
+		} else if strings.ToLower(proceed) != "yes" {
+			c.ui.Output("Destroying project %q and resources requires confirmation.", project.Project.Name, terminal.WithWarningStyle())
+			return 1
+		}
```
If I read this correctly, if error is not `nil` then we output an error message but continue on and attempt to destroy things regardless if the user input `yes` or anything

catsby replied:
Was this resolved by mistake or am I missing something? The problem appears to still be present. Am I mistaken?

xiaolin-ninja reviewed (commented):
Great work! I wanna throw in an existential discussion though; do we want to hard-delete projects? How do we handle orphaned resources?

paladin-devops committed:
Grammar nit for DestroyProject description.

Co-authored-by: Shirley Xiaolin Xu <34314221+xiaolin-ninja@users.noreply.github.com>

paladin-devops commented on pkg/server/proto/server.proto:
```diff
@@ -93,6 +93,10 @@ service Waypoint {
   rpc ListProjects(google.protobuf.Empty) returns (ListProjectsResponse) {}
 
+  // DestroyProject deletes a project from the database as well as (optionally)
+  // the destruction of all resources created within a project
```
@xiaolin-ninja to add some context here, while testing I noticed that if you destroy a project, then if you re-init the project, then the resources, deployments, builds, etc. are readily visible and "attached" to the project again, as if `project destroy` had never been run.

xiaolin-ninja replied:
Sure, but what happens if an endpoint gets or puts the resources while the project is gone, and validation fails because the project doesn't exist?

paladin-devops replied:
[#3673](https://github.com/hashicorp/waypoint/pull/3673) updates this branch with cascading deletion of projects.

paladin-devops commented on internal/runner/operation_project_destroy.go:
```diff
@@ -0,0 +1,97 @@
+	}
+
+	// Update the project state to indicate that it's being destroyed
+	_, err := client.UpsertProject(ctx,
```
@xiaolin-ninja your mention of polling made me more curious about this. Do you mean that while a project destroy operation is running, that a poll operation could simultaneously be run & re-upsert the project in an ACTIVE state?

xiaolin-ninja replied:
Yup that's what I'm thinking.

briancain replied:
Polling shouldn't run on a non-existent project if you delete a project from the db. Project messages have a `DataSourcePoll.Enabled` field that will switch the poller on and off that it checks for when it goes to look at all current projects the poller should queue a message on.

paladin-devops committed:
Add test for destroying a project.

paladin-devops committed:
Return early if error when getting user input for project destroy CLI.

paladin-devops committed:
Fix name of project in project destroy test.

paladin-devops committed:
Project destroy description update.

paladin-devops committed:
Reset state of project back to ACTIVE if destroying a project fails.

paladin-devops committed:
Update project destroy docs.

xiaolin-ninja commented on pkg/serverhandler/handlertest/test_service_project.go:
```diff
@@ -90,6 +90,24 @@ func TestServiceProject(t *testing.T, factory Factory) {
+			require.NoError(err)
+
+			resp, err := client.GetProject(ctx, &pb.GetProjectRequest{
+				Project: &pb.Ref_Project{Project: "not-found"},
```
We should be getting the project with ID "example" to test that project "example" was successfully destroyed.

We should also show two different tests, one that shows resources are not deleted with the destroy resources flag, and one that shows they are successfully deleted without the flag.

paladin-devops replied:
@xiaolin-ninja added! I had to implement it in the `internal/runner` package as the option to skip destroying resources is available when creating the project destroy job only, not here in `service_project`.

paladin-devops committed:
Add delete function to appOperation, to delete an operation record by reference.

paladin-devops committed:
Enable deletion of builds, artifacts, deployments, releases and status reports.

paladin-devops committed:
Update ProjectDelete in bolt to delete other resources associated with a project.

paladin-devops committed:
Revert addition of delete functionality for builds, deployments, artifacts, releases and status reports.

paladin-devops committed:
Commit DB appOperation deletion to in memory DB after bolt DB update.

paladin-devops committed:
Fix release deletion during project DB cleanup. Add workspace deletion during project DB cleanup.

paladin-devops committed:
Add test for app operation deletion.

cicoyle committed:
rebase + update protos

paladin-devops committed:
For project delete, refactor reads into separate transaction. Refactor writes into individual transactions due to bolt DB lock. Add test for creating/deleting a project with a build.

paladin-devops committed:
Regenerate protos after rebase.

paladin-devops committed:
Multiple fixes in project delete bolt DB.

Updated call to delete pipeline to correct function. Updated when the in-memory transaction is set during project delete - moved to be instantiated just before the deletion of the project. Temporarily removed trigger deletion. Updated project test to verify deletion of builds, artifacts, deployments, releases and status reports.

paladin-devops committed:
Tidy.

paladin-devops committed:
Fix workspace deletion function.

Workspace name needs to be converted from uint to string before being deleted from the in-memory database.

paladin-devops committed:
Add deletion of workspaces and triggers to project delete in boltdb.

paladin-devops committed:
Fix workspace deletion on project delete and add to test.

paladin-devops committed:
Add test for deleting triggers.

paladin-devops committed:
Fix name of project in project test for bolt, & fix pipeline get after deletion of project.

paladin-devops committed:
Add key to field of pipeline owner in project test.

paladin-devops committed:
Update rules of workspace deletion on project delete. Fix deletion of triggers.

paladin-devops committed:
Update project delete in boltdb to unset all project and app-scoped config, and update test.

paladin-devops committed:
Add test for deletion of a workspace with a project delete.

paladin-devops committed:
Add test for executing the project destroy runner operation.

paladin-devops committed:
Regenerate protos.

paladin-devops committed:
Update display of destroyed resources to occur only if some resources weren't destroyed.

paladin-devops committed:
Get the latest changes from plugin SDK for setting destroyed/declared resources.

paladin-devops committed:
Check error directly after return of ProjectDelete in test.

paladin-devops committed:
Consolidate declarations of vars in boltdb project delete.

paladin-devops committed:
Fix getting which app operations to delete during project delete.

Prior to this commit, the slices of operations to be deleted were being overwritten with each app. This commit updates the behavior to append them to a slice to be deleted later, so the operations for all apps in a project are retrieved and then deleted. A test is added for this behavior as well.

paladin-devops committed:
Add test for attempting deletion of non-existent operation.

catsby commented on .changelog/3626.txt:
````diff
@@ -0,0 +1,3 @@
+```release-note:feature
+core: Add ability for projects to be deleted from the database with their resources
````
```suggestion
core: Introduce a new CLI command `project destroy` to delete projects in Waypoint and destroy their associated resources
```

catsby commented on go.mod:
```diff
@@ -66,7 +66,7 @@ require (
 	github.com/imdario/mergo v0.3.12
 	github.com/improbable-eng/grpc-web v0.13.0
-	github.com/kevinburke/go-bindata v3.22.0+incompatible
+	github.com/kevinburke/go-bindata v3.23.0+incompatible
```
Is there a reason for this update?

paladin-devops replied:
Whilst setting up the env for OSS Waypoint, I advised @cicoyle to run `go get github.com/kevinburke/go-bindata`, which resulted in the latest version being pulled. It looks like there were a few inconsequential updates between 3.22 and 3.23, so this could be rolled back if updating leads to issues.

krantzinator replied:
Interestingly, we don't appear to specify a version when pulling this for our release build https://github.com/hashicorp/waypoint-releases/blob/70d8ac07d4838b8b90d81bf5583ef5ad988e2648/.circleci/config.yml#L330

catsby commented on internal/cli/job_inspect.go:
```diff
@@ -112,6 +112,8 @@ func (c *JobInspectCommand) Run(args []string) int {
 	case *pb.Job_PipelineStep:
 		op = "PipelineStep"
+	case *pb.Job_DestroyProject:
+		op = "DestroyProject"
```
I believe similar changes need to be made in [`cli/task_list.go`](https://github.com/hashicorp/waypoint/blob/420f95a7f08178ed88b5ce549fb04b45b443e1fd/internal/cli/task_list.go#L102) and [`cli/task_inspect.go`](https://github.com/hashicorp/waypoint/blob/420f95a7f08178ed88b5ce549fb04b45b443e1fd/internal/cli/task_inspect.go#L169), unfortunately :/

catsby commented on internal/cli/project_destroy.go:
```diff
@@ -0,0 +1,102 @@
+	return formatHelp(`
+Usage: waypoint project destroy [options]
+
+  Delete the project and all resources created for all apps in the project, within
+  each app's plugin platform.
```
```suggestion
  Delete the project and all resources created, for all applications in the project
```

I'm not totally sure what "each app's plugin platform" means

paladin-devops replied:
Updated to make it a bit more clear! I removed the mention of plugin, and just noted that the resources created in each platform are deleted.

catsby commented on internal/runner/operation_project_destroy.go:
```diff
@@ -0,0 +1,106 @@
+			Project: destroyProjectOp.DestroyProject.Project.Name,
+		},
+	})
+	// If there is an error destroying a project, reset it's state to ACTIVE
```
```suggestion
	// If there is an error destroying a project, reset its state to ACTIVE
```

catsby reviewed (changes requested):
👍 I think the case checking needs to be added in task list/inspect, the rest is minor feedback

paladin-devops committed:
Fix spelling in app operation test.

Co-authored-by: Shirley Xiaolin Xu <34314221+xiaolin-ninja@users.noreply.github.com>

paladin-devops committed:
Update .changelog/3626.txt

Co-authored-by: Clint <catsby@users.noreply.github.com>

paladin-devops committed:
Add debug logs for deletion of other app operations.

paladin-devops committed:
Delete project-associated records in one transaction.

This commit enables app operations to be deleted with a new function 'delete' that accepts a DB transaction as an input parameter. This allows the delete logic for each of them to be called during an existing transaction without causing a deadlock.

Furthermore, non-app operation records, like pipelines or triggers, are now similarly called, in that the delete function for each, which allows a transaction to be passed as an input parameter, is used.

paladin-devops committed:
Comment grammar fix.

Co-authored-by: Clint <catsby@users.noreply.github.com>

paladin-devops committed:
Update task inspection commands to account for new DestroyProject job type.

paladin-devops committed:
Update project destroy CLI description.

briancain reviewed (changes requested):
Since there are dependent PRs that haven't been merged up into this branch I'll hold off on my full testing, but I caught these while testing this branch.

briancain commented on internal/client/operation.go:
```diff
@@ -28,6 +28,21 @@ func (c *Project) Validate(ctx context.Context, op *pb.Job_ValidateOp) (*pb.Job_
+
+	// Execute it
+	result, err := c.doJob(ctx, job, c.UI)
+	return result.ProjectDestroy, err
```
I think this means there was also an issue with the `ProjectDestroy` job. It likely was returning an error from the operation but the job result is nil so it panic'd

paladin-devops committed:
App operation updates.

This commit updates the deletion of app operations to also be deleted from the in-memory database. Furthermore, when an operation is deleted, its sequence # is decremented. This is necessary because if a project is re-initialized after deletion, new operations would have a sequence # set to the sequence # prior to the project's deletion, plus one.

In order to do this, the app and project name for each operation is required. Thus, the full proto is passed to the delete function for each app operation, instead of just a Ref_Operation.

paladin-devops committed:
Removes unused Delete functions for app operations from state interface, and their implementations.

paladin-devops committed:
Delete hostnames after destroying deployments during the project destroy operation.

paladin-devops committed:
Change name of project and app in sequence test.

paladin-devops committed:
Set error in client operation for project destroy before returning.

paladin-devops committed:
Warn user to not supply arguments for project destroy CLI.

paladin-devops committed:
Fix error with project destroy operation.

Update test for operation project destroy to include testing deletion of a deployment from the DB.

briancain commented on go.mod:
```diff
@@ -334,7 +334,7 @@ require (
 	github.com/hashicorp/go-grpc-net-conn v0.0.0-20220321172933-7ab38178cb90
 	github.com/hashicorp/opaqueany v0.0.0-20220321170339-a5c6ff5bb0ec
-	github.com/hashicorp/waypoint-plugin-sdk v0.0.0-20220502215818-69a3eeb201d8
+	github.com/hashicorp/waypoint-plugin-sdk v0.0.0-20220815171738-0c7040f14e61
```
Don't forget to update this once the sdk PR is merged!

briancain commented on pkg/server/proto/server.proto:
```diff
@@ -3500,6 +3547,8 @@ message Deployment {
   // Resources that this deployment has created or manages.
   repeated DeclaredResource declared_resources = 19;
 
+  repeated DestroyedResource destroyed_resources = 21;
```
This could use a comment too

briancain reviewed (changes requested):
Mostly minor comments from me, but looks great! Lots of work went into this one, nice job! 👏🏻

briancain commented on internal/core/app_deploy_destroy.go:
```diff
@@ -114,11 +118,30 @@ func (a *App) destroyDeploy(
+
+	destroyProto, ok := destroyment.(*pb.Deployment)
+	if !ok {
+		return err
```
I think we have to write our own err here

briancain commented on internal/core/app_deploy_destroy.go:
```diff
@@ -114,11 +118,30 @@ func (a *App) destroyDeploy(
+				a.UI.Output("- "+resource.Name, terminal.WithSuccessStyle())
+			}
+		}
+	}
```
It might be nice to build up these strings in the loops and do a single `a.UI.Output` once rather than for each entry

xiaolin-ninja commented on internal/server/boltdbstate/project.go:
```diff
@@ -71,15 +71,188 @@ func (s *State) ProjectGet(ref *pb.Ref_Project) (*pb.Project, error) {
+			}
+		}
+
+		// delete pipelines for project
```
We should delete pipeline runs too!

briancain commented on internal/server/boltdbstate/project_test.go:
```diff
@@ -0,0 +1,650 @@
+			Workspace: &pb.Ref_Workspace{Workspace: "default"},
+		}))
+
+		// Verify that the app operation sequence is 1, since it is the 1st operation for the new (albeit
```
Maybe add 2 more deployment sequences so we can test more than 1 sequence?

briancain commented on pkg/server/proto/server.proto:
```diff
@@ -2311,6 +2346,14 @@ message Job {
+  // as well as (optionally) the destruction of all resources created within
+  // a project
+  message DestroyProjectOp {
+    Project project = 1;
```
This should probably be a `Ref.Project`

paladin-devops committed:
Update deploy destroy op to fail if the proto message passed to it isn't a Deployment proto.

paladin-devops committed:
Update operation project destroy test to include test for cascade delete.

paladin-devops committed:
Proto change documentation.

paladin-devops committed:
Create error to be returned if the destroyment proto can't be converted to a deployment during a deploy destroy op.

paladin-devops commented on internal/runner/operation_project_destroy.go:
```diff
@@ -0,0 +1,132 @@
+			for _, workspace := range workspaces.Workspaces {
+				log.Debug("Destroying resources in workspace %s", workspace.Name)
+				// TODO: The destroy operation doesn't currently respect the Workspace
+				// being set on the Job - this should be updated
```
It would be nice to fix this! The behavior seems to be rooted in [runner/operation.go](https://github.com/hashicorp/waypoint/blob/main/internal/runner/operation.go#L292-L303). The workspace being used in the destroyment is derived from the `core.Project` that is set here:

```go
	project, err := core.NewProject(ctx,
		core.WithLogger(log),
		core.WithUI(ui),
		core.WithComponents(factories),
		core.WithClient(r.client),
		core.WithConfig(cfg),
		core.WithDataDir(projDir),
		core.WithLabels(job.Labels),
		core.WithVariables(inputVars),
		core.WithWorkspace(job.Workspace.Workspace),  <--- here
		core.WithJobInfo(jobInfo),
	)
// some other code
	case *pb.Job_DestroyProject:
		return r.executeDestroyProjectOp(ctx, log, job, project)
```

So in a way, my comment isn't accurate - the destroyment ONLY respects the workspace that is set on the job and no others... but in our case that is the project destroy job, which happens to need to loop through all workspaces and delete everything. What this means is that if you ran `project destroy -w=dev`, then all of the deployments in a `dev` workspace _would_ be deleted, but no others. But that top-level workspace flag controls which workspace's deployments are going to be destroyed... need to think more on this one but I think this defines the issue.

paladin-devops commented on internal/server/boltdbstate/project_test.go:
```diff
@@ -0,0 +1,650 @@
+	"testing"
+)
+
+func TestProject(t *testing.T) {
```
![Alt Text](https://c.tenor.com/Qky-vwGLdGkAAAAC/moana-youre-welcome.gif)

paladin-devops committed:
Build up message output for destroyed/remaining resources after a destroy op.

paladin-devops committed:
Add additional app operations after re-initting a project after destroying it during a test.

paladin-devops committed:
Update proto for destroy project to use a Ref_Project proto instead of a full Project proto.

paladin-devops committed:
Update error message for failed proto conversion during project destroy to contain text "error" for test.

paladin-devops committed:
Get latest plugin SDK version.

paladin-devops committed:
Tidy.

paladin-devops committed:
Update project destroy CLI to fail without the project flag.

paladin-devops committed:
Cycle through workspaces during project destroy operation.

This commit adds a new function to the project type in core to create a copy of a project with a different workspace. The project destroy operation requires this because when the workspace is set by the runner, executing a destroy operation for an app is only possible in that workspace, regardless of the workspace proto that is set on the job for a destroy operation.

The Waypoint config is required to be plumbed through because the config is needed to re-init a project when we get a copy of it, so we provide it to the project destroy operation from the runner.

evanphx commented on internal/core/project.go:
```diff
@@ -245,6 +245,61 @@ func (p *Project) Ref() *pb.Ref_Project {
+// The project's config is required to be passed in because the Config
+// option is not set on a project, so we can't reference it directly.
+// Getters for other project fields are used here to limit their exposure.
+func (p *Project) Copy(ctx context.Context, workspace string, projConfig *config.Config) (*Project, error) {
```
I'd probably call this `InWorkspace` rather than `Copy` to convey better why it would be used.

paladin-devops committed:
Update Copy function name in core/project to InWorkspace.

paladin-devops commented:
Since making the change to `internal/core`, and the corresponding change to `internal/runner` to make possible the destruction of deployments in _all_ workspaces, the tests I wrote for the runner have broken. Will be looking to see what mock data is required for that test to pass; but aside from that, all feedback up to this point has been addressed!

cicoyle committed:
fix bug if server doesn't have URL service enabled

paladin-devops committed:
Fix broken operation project destroy tests.

paladin-devops committed:
Rebase & regen protos. Use new ref pipeline ID for deleting pipelines during project delete for bolt.

paladin-devops committed:
Rebase & regenerate protos.

krantzinator commented on .changelog/3626.txt:
````diff
@@ -0,0 +1,3 @@
+```release-note:feature
+core: Introduce a new CLI command `project destroy` to delete projects in Waypoint and destroy their associated resources
````
This is important enough that you can highlight the changelog item like how Brian did the Pipelines changelog here https://github.com/hashicorp/waypoint/pull/3777/files

krantzinator reviewed (approved):
If no one else can look at this in the next 20 minutes, go ahead and merge! This has gone through multiple feedback rounds from almost everyone on the team, and has been an extremely thorough process. Because it's definitely going in our release, if it causes regressions those will be found in our regression testing over the next few days and should be handled separately at this point.
💯 🔥 
Edit: per convo in Slack, ensure this error is fixed:
```
Unlinked pages found in the ../content/commands directory.
```

paladin-devops committed:
Update commands nav data file for project destroy command.

paladin-devops committed:
Update project destroy changelog.

paladin-devops committed:
Add remaining mdx files for project destroy command.
